spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/redis/TraceRedisAutoConfiguration.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.instrument.redis;
+
+import brave.Tracing;
+import io.lettuce.core.resource.ClientResources;
+import io.lettuce.core.tracing.BraveTracing;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.cloud.sleuth.autoconfig.TraceAutoConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * {@link org.springframework.boot.autoconfigure.EnableAutoConfiguration
+ * Auto-configuration} enables Redis span information propagation.
+ *
+ * @author Chao Chang
+ * @since 2.2.0
+ */
+@Configuration
+@OnRedisEnabled
+@ConditionalOnBean(Tracing.class)
```

Comment:
It would make sense to add also `@ConditionalOnBean(ClientResources.class)` to ensure a bean is available. One could exclude `RedisAutoConfiguration`. The condition would also report to Spring Boot's conditional reporter and improve the debugging experience.

Reply:
I added `@ConditionalOnBean(ClientResources.class)` to `LettuceConfig` class because this class is the configuration intended for Lettuce, while `TraceRedisAutoConfiguration` is intended for the configuration for all Redis clients, say, Jedis in the future.

Reply:
It's not just about a classpath dependency but also to check whether there is a `ClientResources` bean.

Reply:
Now I added `@ConditionalOnBean(value = { Tracing.class, ClientResources.class })` to `TraceRedisAutoConfiguration` class, making  `TraceRedisAutoConfiguration` be configured only in case of there is a `ClientResources ` bean, is this what advised to do?

Reply:
Yes, that's exactly how it is supposed to work.